src/main/java/cuchaz/enigma/network/packet/PacketRegistry.java

```diff
@@ -0,0 +1,61 @@
+package cuchaz.enigma.network.packet;
+
+import cuchaz.enigma.gui.GuiController;
+import cuchaz.enigma.network.ServerPacketHandler;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class PacketRegistry {
+
+	private static final Map<Class<? extends Packet<ServerPacketHandler>>, Integer> c2sPacketIds = new HashMap<>();
+	private static final Map<Integer, Supplier<? extends Packet<ServerPacketHandler>>> c2sPacketCreators = new HashMap<>();
+	private static final Map<Class<? extends Packet<GuiController>>, Integer> s2cPacketIds = new HashMap<>();
+	private static final Map<Integer, Supplier<? extends Packet<GuiController>>> s2cPacketCreators = new HashMap<>();
+
+	private static <T extends Packet<ServerPacketHandler>> void registerC2S(int id, Class<T> clazz, Supplier<T> creator) {
+		c2sPacketIds.put(clazz, id);
+		c2sPacketCreators.put(id, creator);
+	}
+
+	private static <T extends Packet<GuiController>> void registerS2C(int id, Class<T> clazz, Supplier<T> creator) {
+		s2cPacketIds.put(clazz, id);
+		s2cPacketCreators.put(id, creator);
+	}
+
+	static {
+		registerC2S(0, LoginC2SPacket.class, LoginC2SPacket::new);
```

Comment:
do we need int packet ids? Ive always hated them.

The classname might do tbh

Reply:
Class names can change, though. Int packet IDs are nice and compact, if well abstracted away.

Reply:
Yeah class names are not the answer, especially since we may have other clients implementing the protocol. They could still be short string IDs, but like asie says, I don't really see the problem with ints if they are well encapsulated.

Reply:
> Class names can change

Yep, using the class name would essentially prevent refactoring (while keeping compatibility with previous versions).

Reply:
I think the best thing would be a `PacketType` enum, and use .ordinal() and byOrdinal() when writing and reading.

Reply:
That sounds like an unnecessary layer of abstraction to me, and ordinals can also break. Remember the protocol needs to be well-defined so that other clients (such as maybe Cloak at some point) can connect.

Reply:
Abstracting meaningless numbers into meaningful objects is exactly the reason enums were created.

Java guarantees the ordinals of enum values start at 0 and are in the order the values are declared, because converting to and from ordinals is part of the way enums are intended to be used.

Reply:
But if you want to remove a packet, then all the other IDs shift if you use an enum. But if you are worried about a few integer literals that are only used once, then you can always convert the ID map to a list to the same effect.

Reply:
the big thing is that network protocols tend to change, and using implicit IDs doesn't respond well to changes